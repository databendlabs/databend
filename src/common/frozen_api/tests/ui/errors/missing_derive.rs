// Copyright 2021 Datafuse Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use databend_common_frozen_api::{frozen_api, FrozenAPI};

// Layer 1: Has FrozenAPI derive - OK
#[derive(FrozenAPI)]
struct Config {
    host: String,
    port: u16,
}

// Layer 2: Missing FrozenAPI derive - BROKEN LINK
struct BadService {
    config: Config,  // Config is OK
    timeout: u32,
}

// Layer 3: Should fail because BadService in the chain lacks FrozenAPI
#[frozen_api("dummy")]
#[derive(FrozenAPI)]
struct Application {
    service: BadService,  // ERROR: BadService missing __FROZEN_API_HASH
    version: String,
}

fn main() {}