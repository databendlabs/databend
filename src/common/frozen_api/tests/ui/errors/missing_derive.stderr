error: ⚠️  BREAKING CHANGE DETECTED: 'Application' modified (or its FrozenAPI dependencies)
       Expected: dummy | Actual: dbfc66fc
       Fix: Update hash to 'dbfc66fc' if change is intentional
       Impact: Serialization compatibility may break
  --> tests/ui/errors/missing_derive.rs:31:1
   |
31 | #[frozen_api("dummy")]
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `frozen_api` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no associated item named `__FROZEN_API_HASH` found for struct `BadService` in the current scope
  --> tests/ui/errors/missing_derive.rs:31:1
   |
25 | struct BadService {
   | ----------------- associated item `__FROZEN_API_HASH` not found for this struct
...
31 | #[frozen_api("dummy")]
   | ^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `BadService`
   |
   = note: this error originates in the attribute macro `frozen_api` (in Nightly builds, run with -Z macro-backtrace for more info)
